### ISSUE TITLE
Webapp gets incomplete data from the MariaDB for Jenkins

### ISSUE THREAD

**Issue by ian-jarrett** [Juniper/nita-webapp#21]
Tried to build a simple syslog config using this JSON data:
```
{"syslog": [{"loghost": "100.123.0.16","facility": "daemon","level":"info"},{"loghost": "100.123.0.16","facility": "any","level": "warning"}]}
```
Run write_yaml_files.py from within the jenkins container to generate ./group_vars/spines.yaml:
```
syslog:
- loghost: 100.123.0.16
  facility: daemon
  level: info
- loghost: 100.123.0.16
  facility: any
  level: warning
```
And use this J2 template to build the Junos config:
```
system {
  syslog {
    {% for host in syslog %}
      host {{ host.loghost }} {
        {{ host.facility }} {{ host.level }};
      }
    {% endfor %}
  }
}
```
If running from the shell, the build process works correctly. If running from the Webapp, Ansible fails with this error:
```
AnsibleUndefinedVariable: 'ansible.parsing.yaml.objects.AnsibleUnicode object' has no attribute 'loghost'"
```
This error is because the YAML file was not fully built by the "Build" job. However, the Excel data is all correctly stored in the "ngcn_worksheets" table inside the "Sites" database (running in the MariaDB container). So somehow, this data is not getting extracted correctly by the webapp when it runs the Jenkins job.

**Comment by wildsubnet** [comment]
Confirming issue. MariaDB has both rows of data but the data.json that is generated when the ansible-playbook is run  run by jenkins only gets the last row of data. 

Tested against ansible 2.10 as well just in case it is some oddball issue with ansible. Same result.

**Comment by hribeiro-jnpr** [comment]
Yes that was my understanding as well, pretty sure the issue is around views.py where the webapp gets the info from mariadb and generates the data.json.

**Comment by wildsubnet** [comment]
Confirmed based on Debug info that data is being truncated in views.py code. Output in log shows data set at line 207 but my 924 the extra rows are gone. Querying Maria DB directly using the same selector/order by as the python code returns the data as expected. So something from the data retrieval into json file is losing fidelity. This might be being done by built-in functions so maybe a bug upstream (json library maybe?)

**Comment by wildsubnet** [comment]
Potential workaround for now...for multi-line entries like syslog you can create a second tab called anything (I chose syslog2) and the yaml files get created properly and the data is there in the debug log for the data.json file being pulled by views.py. 

Generated Yaml file on test data: 
---
OS_version: 18.1R3-S5.3
healthbot_device_group: dc1-spines
loopback_ip: 10.30.100.3
management_interface:
  int: em0
  ip: 100.123.24.3
  mask: 16
syslog:
  facility: any
  level: warning
  loghost: 100.123.0.17
syslog2:
  facility: daemon
  level: info
  loghost: 100.123.0.17

Ansible still doesn't like "loghost" as a variable name, for some reason. I suspect it might be unrelated to this.

**Comment by ian-jarrett** [comment]
Thanks Matt, I've tried your workaround and I can confirm that it works, thank you very much. We can caveat this behaviour in the docs for now and move on. Whether it's an ER or a bug I don't know, I'll defer to you since you spent longer looking into the code than I did :-)

**Comment by aburston** [comment]
Doing a basic test on the example data definitely shows that this problem is not caused by yaml2xls:

test.yaml
```
syslog:
- loghost: 100.123.0.16
  facility: daemon
  level: info
- loghost: 100.123.0.16
  facility: any
  level: warning
```
`yaml2xls.py test.yaml test.xlsx` produces a file `test.xlsx`
then you can convert this file back to yaml by doing
`xls2yaml.py test.xlsx tmp` and that produces:

`tmp/test.yaml`
```
---
syslog:
- loghost: 100.123.0.16
  facility: daemon
  level: info
- loghost: 100.123.0.16
  facility: any
  level: warning
```
Which has slightly different ordering but nothing important has changed.

**Comment by wildsubnet** [comment]
This seems to be resolved by an undocumented feature in NITA. If we change the column header on syslog to be host+.facility and host+.level the views.py subroutines use ordered_dicts as fields instead of strings. So we go from this:
 
('syslog', OrderedDict([('loghost', '100.123.0.17'), ('facility', 'daemon'), ('level', 'info')]))])),
 
To this:
 
('syslog', OrderedDict([('loghost', '100.123.0.17'), ('host', [OrderedDict([('facility', 'any'), ('level', 'warning')]), OrderedDict([('facility', 'daemon'), ('level', 'info')])])]))])),
 
Then after that I believe it should just be updating the j2 template to iterate through each element like the evpn_port template does in the examples. If you look at the example files for the data center use case evpn_port tab there is "+" on the tab name too. I assume it’s doing some sort of larger grouping that may not matter for the syslog example (I’m almost sure that’s it honestly). However, this worked without the tab name being "syslog+". Something to explore further while working on developer's guide.

**Comment by aburston** [comment]
This sounds like a documentation bug against nita-yaml-to-excel
